File: OJ/leet206/main.py

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList1(self, head: ListNode) -> ListNode:
        def rev(curr, prev):
            if curr is None: return prev
            nextNode, curr.next = curr.next, prev
            return rev(nextNode, curr)
        return rev(head, None)

    def reverseList(self, head: ListNode) -> ListNode:
        prev, curr = None, head
        while curr:
            nextNode, curr.next = curr.next, prev 
            prev, curr = curr, nextNode
        return prev 

def printList(node):
    while node:
        print(node.val)
        node = node.next       
            

head = ListNode(-1)
p = head 
for i in range(3):
    p.next = ListNode(i)
    p = p.next

printList(head)
rev = Solution().reverseList(head)
printList(rev)

File: OJ/leet206/main.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }

  fn print(&self) {
      
  }
}

type Link = Option<Box<ListNode>>;
impl Solution {
    pub fn reverse_list(head: Link) -> Link {
        // 
        fn rev(cur: Link, prev: Link) -> Link {
            if let Some(mut p) = cur {
                let nextNode = p.next;
                p.next = prev;
                return rev(nextNode, Some(p));
            }
            else { return prev; }
        }
        return rev(head, None);
    }
}

fn main() {
    // let sol = Solution;
    let head = ListNode { 0i32, None};
    // head.next = new(1);
    let bx = Box::new(head);
    println!("bx.val = {:?}", bx.val);
    println!("head.next = {:?}", head.next);
    
    // sol.reverse_list()
}